FEATURE: Make auth_redirect param options on user_api_keys

cfitz opened the pull request:
This is a possible solution for https://meta.discourse.org/t/user-api-keys-specification/48536/19
This allows for user-api-key requests to not require a redirect url.
Instead, the encypted payload will just be displayed after creation  ( which can be copied
pasted into an env for a CLI, for example  )

vsoch commented:
Huge +1 to this addition! The first working example (also to be merged soon!) is via the helpme client :point_right: https://github.com/vsoch/helpme/pull/34 when you write:

```
helpme discourse https://stage.neurostars.org
```
(or some other board) it will walk you through posting an issue, meaning a title, text, optional asciinema and environment, and then post without leaving the command line! This PR here allows for the generation of the user token on the first try (when the user doesn't have one yet).

SamSaffron commented:
Just rendering the payload seems somewhat surprising, I would expect some sort of instructions on the final page, eg:

> We just generated a new user API key for you, please paste the following key into your application:
>
> abc12312ab12312312321

vsoch commented:
The generation of the token happens by way of the client, and the user needs to copy paste the payload into the terminal. This PR redirects the user to the page with the payload, and the token generation needs to happen by the client so we can use their credentials files. The instructions are also clear when the interaction is happening. It also only needs to happen once, so it's not too much of a hassle.

SamSaffron commented:
I follow, but a real page here is very beneficial. It is then clear you are on site cause the header is still there, and there are a more instructions to go with cause we don't simply render a white page with a code. 

It is a very simple page rendering this with some canned message and the no-ember layout.

cfitz commented:
Adding a view makes sense to me. But also, if the request is expecting a JSON response, should we return JSON? I added both, but let me know if I should just return the html version.

Also I added a new i18n entry to the server.en.yml. Is there a process I should follow to add these to the other translations?

Lastly, should I be squashing my commits?

vsoch commented:
If the user needs to physically click to approve, how can we expect Json? Would they approve, leave the window open, then press enter (and then the client does a GET to the page with json?) Is that a better interaction than copy paste?

cfitz commented:
I think the login and the approval will still need to be done via the browser. The only change is if a user post's to user-api-key.json they'll get a JSON response with the payload and the instructions message instead of the html page.
Biut, yeah, that might be unnecessary.

vsoch commented:
Is this change live with stage.neurostars.org? If so I'd like to run it through testing again to see the new interaction (and try out the json).

cfitz commented:
@vsoch just updated it. try it out and see if that works for you.

SamSaffron commented on app/controllers/user_api_keys_controller.rb:
```diff
@@ -76,17 +77,24 @@ def create
-    redirect_to "#{params[:auth_redirect]}?payload=#{CGI.escape(payload)}"
+    @payload = Base64.encode64(public_key.public_encrypt(@payload))
+
+    (redirect_to("#{params[:auth_redirect]}?payload=#{CGI.escape(@payload)}") && return) if params[:auth_redirect]
```
style wise ... the && return is just so easy to miss ... 

if params[:auth_redirect]
   ... do one thing
else
   ... do another thing
end

SamSaffron commented on app/controllers/user_api_keys_controller.rb:
```diff
@@ -51,22 +51,23 @@ def create
     require_params
 
-    unless SiteSetting.allowed_user_api_auth_redirects
+    if params.key?(:auth_redirect) && SiteSetting.allowed_user_api_auth_redirects
```
yeah ... unless is just so hard to reason about ... agree with this change and sorry about using it there in the first place :)

SamSaffron commented:
one minor comment and we are good to merge 🎊

vsoch commented:
Testing now.

vsoch commented:
hey @cfitz the page is totally blank
![image](https://user-images.githubusercontent.com/814322/50613820-df001880-0eac-11e9-803d-a930d6d642a7.png)

vsoch commented:
Here is the actual page, looks like some kind of blank 404? 
[notfound.html.txt](https://github.com/discourse/discourse/files/2722161/notfound.html.txt)

vsoch commented:
Does the flow need to change at all?

vsoch commented:
I'll try the .json version and see if that shows anything.

vsoch commented:
It's not clear to me if I'm supposed to do user-api-key.json/new, or user-api-key/new.json - both trigger an error. Here is doing:

```python
    url = (board + "/user-api-key.json/new?scopes=write&application_name=HelpMe&public_key=" + 
           self.public_key.replace("'", "") + 
           "&client_id=" + self.client_id +
           "&nonce=" + nonce )
```

![image](https://user-images.githubusercontent.com/814322/50614273-3fdc2080-0eae-11e9-9292-e1bd586360c5.png)

And then doing the other way:

```python
    url = (board + "/user-api-key/new.json?scopes=write&application_name=HelpMe&public_key=" + 
           self.public_key.replace("'", "") + 
           "&client_id=" + self.client_id +
           "&nonce=" + nonce )
```
![image](https://user-images.githubusercontent.com/814322/50614415-b1b46a00-0eae-11e9-9450-97ac0452a7f9.png)

And if I leave as it was before (not asking for json) I get an empty page. Let me know what to try!

cfitz commented:
@vsoch  Sorry, that's my bad! I realized the last git pull on the stage server actually errored because the git history diverged. I just reset it and pulled. I think it should be working now.

 Sorry, I think I confused things about the JSON handler...the test suite POSTs directly to the user_api_keys controller.  The /user-api-key/new ( without the .json ) action is the form that the user needs to approve.

vsoch commented:
okay, I will try again (with the original flow)

vsoch commented:
Oh it's lovely! So much better!

![image](https://user-images.githubusercontent.com/814322/50615323-0e655400-0eb2-11e9-9fd2-5d3d4d33bb8b.png)

vsoch commented:
Still testing though :) Hang tight!

vsoch commented:
okay it looks good! Here is the post that I just made:

https://stage.neurostars.org/t/why-are-jelly-beans-so-funky-looking/29

And it has it's own asciinema, but I also recorded a second one to see the client interaction (so @SamSaffron  can see too!)

https://asciinema.org/a/219484?speed=2

vsoch commented:
@SamSaffron did you have any other issues we need to work on?

SamSaffron commented:
Looks good to me! thanks heaps!
